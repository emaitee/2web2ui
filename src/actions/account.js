import sparkpostApiRequest from 'src/actions/helpers/sparkpostApiRequest';

export function fetch(params = {}) {
  return sparkpostApiRequest({
    type: 'FETCH_ACCOUNT',
    meta: {
      method: 'GET',
      url: '/account',
      params
    }
  });
}

export function getPlans() {
  return sparkpostApiRequest({
    type: 'GET_PLANS',
    meta: {
      method: 'GET',
      url: '/account/plans'
    }
  });
}

export function update(data) {
  return sparkpostApiRequest({
    type: 'UPDATE_ACCOUNT',
    meta: {
      method: 'PUT',
      url: '/account',
      data
    }
  });
}


export function register(data) {

  return sparkpostApiRequest({
    type: 'CREATE_ACCOUNT',
    meta: {
      method: 'POST',
      url: '/account',
      data
    }
  });
}
