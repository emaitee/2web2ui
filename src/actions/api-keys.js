import { formatKeyForRequest } from './helpers/api-keys';
import sparkpostApiRequest from './helpers/sparkpostApiRequest';
import { showAlert } from './globalAlert';

export function createApiKey(key) {
  return (dispatch, getState) =>
    dispatch(
      sparkpostApiRequest({
        type: 'CREATE_API_KEY',
        meta: {
          method: 'POST',
          url: '/api-keys',
          ...formatKeyForRequest(key, getState)
        }
      })
    )
      .then(() =>
        dispatch(showAlert({ type: 'success', message: 'API key created' }))
      )
      .catch((err) =>
        dispatch(
          showAlert({
            type: 'error',
            message: 'Could not create API key',
            details: err.message
          })
        )
      );
}

export function deleteApiKey(id) {
  return (dispatch) =>
    dispatch(
      sparkpostApiRequest({
        type: 'DELETE_API_KEY',
        meta: {
          method: 'DELETE',
          url: `/api-keys/${id}`
        }
      })
    )
      .then(() =>
        dispatch(showAlert({ type: 'success', message: 'API key deleted' }))
      )
      .catch((err) =>
        dispatch(
          showAlert({
            type: 'error',
            message: 'Could not delete API key',
            details: err.message
          })
        )
      );
}

export function updateApiKey(id, key) {
  return (dispatch, getState) =>
    dispatch(
      sparkpostApiRequest({
        type: 'UPDATE_API_KEY',
        meta: {
          method: 'PUT',
          url: `/api-keys/${id}`,
          ...formatKeyForRequest(key, getState)
        }
      })
    )
      .then(() =>
        dispatch(showAlert({ type: 'success', message: 'API key updated' }))
      )
      .catch((err) =>
        dispatch(
          showAlert({
            type: 'error',
            message: 'Could not update API key',
            details: err.message
          })
        )
      );
}

export function hideNewApiKey() {
  return {
    type: 'HIDE_NEW_API_KEY'
  };
}

export function listApiKeys({ force } = {}) {
  return (dispatch, getState) => {
    if (!force && getState().apiKeys.keysLoaded) {
      return;
    }

    return dispatch(
      sparkpostApiRequest({
        type: 'LIST_API_KEYS',
        meta: {
          method: 'GET',
          url: '/api-keys'
        }
      })
    );
  };
}

export function listGrants({ force } = {}) {
  return (dispatch, getState) => {
    if (!force && getState().apiKeys.grantsLoaded) {
      return;
    }

    return dispatch(
      sparkpostApiRequest({
        type: 'LIST_GRANTS',
        meta: {
          method: 'GET',
          url: '/authenticate/grants'
        }
      })
    );
  };
}
