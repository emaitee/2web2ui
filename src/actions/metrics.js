import moment from 'moment';
import sparkpostApiRequest from 'src/actions/helpers/sparkpostApiRequest';

const apiFormat = 'YYYY-MM-DDTHH:MM';
const defaultParams = () => ({
  from: moment().subtract(30, 'days').format(apiFormat),
  metrics: 'count_targeted'
});

export function fetch({ type = 'FETCH_METRICS', path, params = {}}) {
  return sparkpostApiRequest({
    type,
    meta: {
      method: 'GET',
      url: `/metrics/${path}`,
      params: {
        ...defaultParams(),
        ...params
      }
    }
  });
}

export function fetchMetricsDomains(params = {}) {
  const type = 'FETCH_METRICS_DOMAINS';
  const path = 'domains';
  return fetch({ type, path, params });
}

export function fetchMetricsCampaigns(params = {}) {
  const type = 'FETCH_METRICS_CAMPAIGNS';
  const path = 'campaigns';
  return fetch({ type, path, params });
}

export function fetchMetricsSendingIps(params = {}) {
  const type = 'FETCH_METRICS_SENDING_IPS';
  const path = 'sending-ips';
  return fetch({ type, path, params });
}

export function fetchMetricsIpPools(params = {}) {
  const type = 'FETCH_METRICS_IP_POOLS';
  const path = 'ip-pools';
  return fetch({ type, path, params });
}

export function fetchDeliverability(params = {}) {
  const path = 'deliverability';
  return fetch({ path, params });
}

export function getTimeSeries(params = {}) {
  const path = 'deliverability/time-series';
  return fetch({ path , params });
}
