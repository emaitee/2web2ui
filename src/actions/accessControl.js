import { fetch as fetchAccount } from './account';
import { get as getCurrentUser, getGrants } from './currentUser';

// initialize some state used for access control
export function initializeAccessControl() {
  return (dispatch) => Promise.all([
    dispatch(fetchAccount()),
    dispatch(getCurrentUser())
      .then(({ access_level }) => dispatch(getGrants({ role: access_level })))
  ])
    .then(() => dispatch({ type: 'ACCESS_CONTROL_READY' }))
    .catch((err) => {
    // TODO: log to SENTRY
    // for now, just ignore this error
    });
}
