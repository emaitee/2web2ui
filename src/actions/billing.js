import { formatDataForCors, formatCreateData } from 'src/helpers/billing';
import { fetch as fetchAccount } from './account';
import sparkpostApiRequest from 'src/actions/helpers/sparkpostApiRequest';
import zuoraRequest from 'src/actions/helpers/zuoraRequest';

export function syncSubscription() {
  return sparkpostApiRequest({
    type: 'SYNC_SUBSCRIPTION',
    meta: {
      method: 'POST',
      url: '/account/subscription/check'
    }
  });
}

export function updateSubscription(code) {
  return sparkpostApiRequest({
    type: 'UPDATE_SUBSCRIPTION',
    meta: {
      method: 'PUT',
      url: '/account/subscription',
      data: { code }
    }
  });
}

/**
 * For updating billing info via our API (e.g. contact info)
 * @param {Object} data
 */
export function updateBilling(data) {
  return sparkpostApiRequest({
    type: 'UPDATE_BILLING',
    meta: {
      method: 'PUT',
      url: '/account/subscription',
      data
    }
  });
}

export function cors(context, data = {}) {
  const type = `CORS_${context.toUpperCase().replace('-', '_')}`;
  return sparkpostApiRequest({
    type,
    meta: {
      method: 'POST',
      url: '/account/cors-data',
      params: { context },
      data
    }
  });
}

export function updateCreditCard({ data, token, signature }) {
  return zuoraRequest({
    type: 'ZUORA_UPDATE_CC',
    meta: {
      method: 'POST',
      url: '/payment-methods/credit-cards',
      data,
      headers: { token, signature }
    }
  });
}

export function updateAddons(product, data) {
  return sparkpostApiRequest({
    type: `UPDATE_ADDON_${product.toUpperCase()}`,
    meta: {
      method: 'POST',
      url: `/account/add-ons/${product}`,
      data
    }
  });
}

export function addDedicatedIps(data) {
  return updateAddons('dedicated_ips', data);
}

export function createZuoraAccount({ data, token, signature }) {
  return zuoraRequest({
    type: 'ZUORA_CREATE',
    meta: {
      method: 'POST',
      url: '/accounts',
      data,
      headers: { token, signature }
    }
  });
}

export function billingCreate(values) {
  const { corsData, billingData } = formatDataForCors(values);

  return (dispatch) =>

    // get CORS data for the create account context
    dispatch(cors('create-account', corsData))

      // create the Zuora account
      .then((results) => {
        const { token, signature } = results;
        const data = formatCreateData({ ...results, ...billingData });
        return dispatch(createZuoraAccount({ data, token, signature }));
      })

      // sync our db with new Zuora state
      .then(() => dispatch(syncSubscription()))

      // refetch the account
      .then(() => dispatch(fetchAccount()));
}

export function getBillingCountries() {
  return sparkpostApiRequest({
    type: 'GET_COUNTRIES_BILLING',
    meta: {
      method: 'GET',
      url: '/account/countries',
      params: {
        filter: 'billing'
      }
    }
  });
}
