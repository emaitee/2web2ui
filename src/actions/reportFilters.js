import {
  fetchMetricsDomains,
  fetchMetricsCampaigns,
  fetchMetricsSendingIps,
  fetchMetricsIpPools
} from './metrics';

import { listTemplates } from './templates';
import { list as listSubaccounts } from './subaccounts';
import { list as listSendingDomains } from './sendingDomains';
import { getRelativeDates } from 'src/helpers/date';
import { getQueryFromOptions } from 'src/helpers/metrics';

// array of all lists that need to be re-filtered when time changes
const metricLists = [
  fetchMetricsDomains,
  fetchMetricsCampaigns,
  fetchMetricsSendingIps,
  fetchMetricsIpPools
];

/**
 * Returns a thunk that gets the metrics filter lists to populate cache,
 * plus templates, subaccounts, and sending domains
 *
 * The thunk is a no-op if the cache has already been initialized, i.e.
 * metrics.domains is an array and not undefined
 */
export function initTypeaheadCache(params) {
  return (dispatch, getState) => {
    const { metrics } = getState();
    if (Array.isArray(metrics.domains)) {
      return;
    }
    const requests = metricLists.map((list) => dispatch(list(params)));
    requests.push(dispatch(listTemplates()), dispatch(listSubaccounts()), dispatch(listSendingDomains()));
    return Promise.all(requests);
  };
}

// TODO: rename refreshTypeaheadCache
export function refreshTypeaheadCache(params) {
  return (dispatch) => {
    const requests = metricLists.map((list) => dispatch(list(params)));
    return Promise.all(requests);
  };
}

export function maybeRefreshTypeaheadCache(dispatch, options, updates = {}) {
  // refresh the typeahead cache if the date range has been updated
  const { relativeRange, from, to } = updates;

  if (relativeRange || from || to) {
    const params = getQueryFromOptions({ from: options.from, to: options.to });
    dispatch(refreshTypeaheadCache(params));
  }
}

export function addFilters(payload) {
  return {
    type: 'ADD_FILTERS',
    payload
  };
}

export function removeFilter(payload) {
  return {
    type: 'REMOVE_FILTER',
    payload
  };
}

// refresh the date range
export function refreshReportRange(options) {
  return {
    type: 'REFRESH_REPORT_RANGE',
    payload: options
  };
}

export function refreshRelativeRange(options = {}) {
  const range = { ...options, ...getRelativeDates(options.relativeRange) };

  return {
    type: 'REFRESH_REPORT_RANGE',
    payload: range
  };
}
