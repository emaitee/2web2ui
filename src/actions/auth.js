import { sparkpostLogin } from '../helpers/http';
import authCookie from '../helpers/authCookie';
import { initializeAccessControl } from './accessControl';


export function login(authData) {
  return (dispatch) => {
    dispatch({
      type: 'LOGIN_SUCCESS',
      payload: authData
    });

    dispatch(initializeAccessControl());
  };
}

export function authenticate(username, password, rememberMe = false) {
  // return a thunk
  return (dispatch, getState) => {
    const { loggedIn } = getState().auth;

    if (loggedIn) {
      return;
    }

    dispatch({ type: 'LOGIN_PENDING' });

    sparkpostLogin(username, password, rememberMe)
      .then(({ data = {}} = {}) => {
        const payload = { ...data, username };

        // save auth cookie
        authCookie.save(payload);

        // dispatch login success event
        dispatch(login(payload));
      })
      .catch((err) => {
        const { response = {}} = err;
        const { data = {}} = response;
        const { error_description: errorDescription } = data;

        // TODO: handle a timeout error better

        dispatch({
          type: 'LOGIN_FAIL',
          payload: {
            errorDescription
          }
        });
      });
  };
}

export function confirmPassword(username, password) {
  return (dispatch, getState) => {
    dispatch({ type: 'CONFIRM_PASSWORD' });

    return sparkpostLogin(username, password, false)
      .then(({ data = {}} = {}) => {
        const payload = { ...data, username };

        // dispatch login success event
        dispatch({
          type: 'CONFIRM_PASSWORD_SUCCESS',
          payload
        });
      })
      .catch((err) => {
        const { response = {}} = err;
        const { data = {}} = response;
        const { error_description: errorDescription } = data;
        dispatch({
          type: 'CONFIRM_PASSWORD_FAIL',
          payload: {
            errorDescription
          }
        });
        throw err;
      });
  };
}

export function beginRefresh(refreshToken) {
  return { type: 'ATTEMPTING_REFRESH', payload: refreshToken };
}

export function refresh(access_token, refresh_token) {
  const newCookie = authCookie.merge({ access_token, refresh_token });
  return login(newCookie);
}

export function logout() {
  return (dispatch) => {
    authCookie.remove();
    dispatch({
      type: 'LOGOUT'
    });
  };
}
