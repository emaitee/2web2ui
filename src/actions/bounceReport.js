import { fetchDeliverability, fetchBounceClassifications, fetchBounceReasonsByDomain } from 'src/actions/metrics';
import { maybeRefreshTypeaheadCache, refreshReportRange } from 'src/actions/reportFilters';
import { getQueryFromOptions, buildCommonOptions, getMetricsFromKeys } from 'src/helpers/metrics';
import { getBandTypes, reshapeCategories, formatAggregates } from 'src/helpers/bounce';
import _ from 'lodash';


// Load new metrics for the bounce chart
// Dispatches refreshBounceChart() on completion.
export function refreshBounceChartMetrics(updates = {}) {
  return (dispatch, getState) => {
    const state = getState();

    const bounceMetrics = [
      'count_targeted',
      'count_bounce',
      'count_inband_bounce',
      'count_outofband_bounce'
    ];

    updates.metrics = getMetricsFromKeys(bounceMetrics);

    const options = buildCommonOptions(state.reportFilters, updates);

    maybeRefreshTypeaheadCache(dispatch, options, updates);

    dispatch(refreshReportRange(options));

    // convert new meta data into query param format
    const aggregateParams = _.omit(getQueryFromOptions(options), 'precision');
    const bounceParams = { ...aggregateParams, metrics: 'count_bounce' };

    // get new data
    return dispatch(fetchDeliverability(aggregateParams))
      .then((aggregates) => {
        if (!aggregates[0].count_bounce) {
          return;
        }

        dispatch(fetchBounceClassifications(bounceParams))
          .then((classifications) => dispatch(refreshBounceChart({ aggregates, classifications })));
      });
  };
}

// Refresh the chart with the new data
export function refreshBounceChart({ aggregates, classifications }) {
  const formattedAggregates = formatAggregates(aggregates[0]);
  return {
    type: 'REFRESH_BOUNCE_CHART',
    payload: {
      categories: reshapeCategories(classifications),
      aggregates: formattedAggregates,
      types: getBandTypes(formattedAggregates)
    }
  };
}

// Load new metrics for the bounce reasons table
// Dispatches refreshBounceTable() on completion.
export function refreshBounceTableMetrics(updates = {}) {
  return (dispatch, getState) => {
    const state = getState();
    const options = buildCommonOptions(state.reportFilters, updates);
    const query = getQueryFromOptions(options);
    const baseParams = _.omit(query, 'precision');
    const params = { ...baseParams, metrics: 'count_bounce' };

    return dispatch(fetchBounceReasonsByDomain(params))
      .then((reasons) => {
        dispatch(refreshBounceTable({ reasons }));
      });
  };
}

// Refhres the table with the new data
export function refreshBounceTable({ reasons }) {
  return {
    type: 'REFRESH_BOUNCE_TABLE',
    payload: { reasons }
  };
}

