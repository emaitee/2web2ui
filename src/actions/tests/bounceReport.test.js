import * as bounceReport from '../bounceReport';
import * as metricsActions from 'src/actions/metrics';
import * as metricsHelpers from 'src/helpers/metrics';
import * as reportFilters from 'src/actions/reportFilters';

import moment from 'moment';

jest.mock('../helpers/sparkpostApiRequest', () => jest.fn((a) => a));
jest.mock('src/helpers/bounce');
jest.mock('src/helpers/metrics');
jest.mock('src/actions/metrics');
jest.mock('src/actions/reportFilters');

describe('Action Creator: Bounce Report', () => {

  const from = moment(new Date(1487076708000)).utc().format('YYYY-MM-DDTHH:MM');
  let dispatchMock;
  let getStateMock;
  let stateMock;
  let getRelDatesSpy;

  beforeEach(() => {
    const metrics = 'count_bounce';
    metricsActions.fetchDeliverability = jest.fn(() => [{ count_bounce: 1 }]);
    metricsHelpers.getQueryFromOptions.mockImplementation(() => ({ from, metrics }));

    stateMock = {};

    dispatchMock = jest.fn((a) => Promise.resolve(a));
    getStateMock = jest.fn(() => stateMock);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should dispatch a chart refresh action', async() => {
    const thunk = bounceReport.refreshBounceChartMetrics();
    await thunk(dispatchMock, getStateMock);
    expect(reportFilters.maybeRefreshTypeaheadCache).toHaveBeenCalledTimes(1);
    expect(reportFilters.refreshReportRange).toHaveBeenCalledTimes(1);
    expect(metricsActions.fetchBounceClassifications).toHaveBeenCalledTimes(1);
    expect(metricsActions.fetchBounceClassifications.mock.calls).toMatchSnapshot();

    expect(dispatchMock.mock.calls).toMatchSnapshot();
  });

  it('should not refresh chart nor update bounce classifications if 0 bounces', async() => {
    metricsActions.fetchDeliverability = jest.fn(() => [{ foo: 'bar' }]);
    const thunk = bounceReport.refreshBounceChartMetrics();
    await thunk(dispatchMock, getStateMock);
    expect(metricsActions.fetchDeliverability).toHaveBeenCalledTimes(1);
    expect(metricsActions.fetchDeliverability.mock.calls).toMatchSnapshot();
    expect(metricsActions.fetchBounceClassifications).not.toHaveBeenCalled();

    expect(dispatchMock.mock.calls).toMatchSnapshot();
  });

  it('should dispatch a table refresh action', async() => {
    const updates = {};
    stateMock.reportFilters = {};
    const mockOptions = {};
    metricsHelpers.buildCommonOptions = jest.fn(() => mockOptions);
    metricsHelpers.getQueryFromOptions = jest.fn(() => ({
      precision: 'cool', // verifying that precision gets stripped out in the snapshot below?
      other: 'stuff'
    }));
    const thunk = bounceReport.refreshBounceTableMetrics(updates);
    await thunk(dispatchMock, getStateMock);
    expect(metricsHelpers.buildCommonOptions).toHaveBeenCalledWith(stateMock.reportFilters, updates);
    expect(metricsHelpers.getQueryFromOptions).toHaveBeenCalledWith(mockOptions);

    expect(metricsActions.fetchBounceReasonsByDomain.mock.calls).toMatchSnapshot();
    expect(dispatchMock.mock.calls).toMatchSnapshot();
  });

});
