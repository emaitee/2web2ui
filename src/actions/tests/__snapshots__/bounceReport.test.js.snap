// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action Creator: Bounce Report should dispatch a chart refresh action 1`] = `
Array [
  Array [
    Object {
      "from": "2017-02-14T12:02",
      "metrics": "count_bounce",
    },
  ],
]
`;

exports[`Action Creator: Bounce Report should dispatch a chart refresh action 2`] = `
Array [
  Array [
    undefined,
  ],
  Array [
    Array [
      Object {
        "count_bounce": 1,
      },
    ],
  ],
  Array [
    undefined,
  ],
  Array [
    Object {
      "payload": Object {
        "aggregates": undefined,
        "categories": undefined,
        "types": undefined,
      },
      "type": "REFRESH_BOUNCE_CHART",
    },
  ],
]
`;

exports[`Action Creator: Bounce Report should dispatch a table refresh action 1`] = `
Array [
  Array [
    Object {
      "metrics": "count_bounce",
      "other": "stuff",
    },
  ],
]
`;

exports[`Action Creator: Bounce Report should dispatch a table refresh action 2`] = `
Array [
  Array [
    undefined,
  ],
  Array [
    Object {
      "payload": Object {
        "reasons": undefined,
      },
      "type": "REFRESH_BOUNCE_TABLE",
    },
  ],
]
`;

exports[`Action Creator: Bounce Report should not refresh chart nor update bounce classifications if 0 bounces 1`] = `
Array [
  Array [
    Object {
      "from": "2017-02-14T12:02",
      "metrics": "count_bounce",
    },
  ],
]
`;

exports[`Action Creator: Bounce Report should not refresh chart nor update bounce classifications if 0 bounces 2`] = `
Array [
  Array [
    undefined,
  ],
  Array [
    Array [
      Object {
        "foo": "bar",
      },
    ],
  ],
]
`;
